# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [85]:
import pandas as pd
import sqlite3 as s3
import warnings

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [86]:
go_sales = s3.connect('go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [165]:
product = pd.read_sql_query('SELECT * FROM product', go_sales)
product_type = pd.read_sql_query('SELECT * FROM product_type', go_sales)
product_line = pd.read_sql_query('SELECT * FROM product_line', go_sales)
sales_staff = pd.read_sql_query('SELECT * FROM sales_staff', go_sales)
sales_branch = pd.read_sql_query('SELECT * FROM sales_branch', go_sales)
retailer_site = pd.read_sql_query('SELECT * FROM retailer_site', go_sales)
country = pd.read_sql_query('SELECT * FROM country', go_sales)
order_header = pd.read_sql_query('SELECT * FROM order_header', go_sales)
order_details = pd.read_sql_query('SELECT * FROM order_details', go_sales)
sales_target = pd.read_sql_query('SELECT * FROM sales_targetData', go_sales)
returned_item = pd.read_sql_query('SELECT * FROM returned_item', go_sales)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [102]:
cost_range = (product['PRODUCTION_COST'].astype(float) < 100) & (product['PRODUCTION_COST'].astype(float) > 50)
pd.DataFrame(product.loc[cost_range, ['PRODUCT_NAME', 'PRODUCTION_COST']])

PRODUCT_NAME PRODUCTION_COST
13                Blue Steel Putter              60
14            Blue Steel Max Putter            81.8
17              Course Pro Golf Bag              80
24                  Hibernator Lite              60
25                       Hibernator              86
28  Hibernator Self - Inflating Mat           78.55
31              Hibernator Camp Cot           65.33
32     Canyon Mule Climber Backpack            62.5
56          Granite Climbing Helmet           52.86
58            Husky Harness Extreme              55
69                   Granite Hammer           57.13
71        TrailChef Deluxe Cook Set           85.11
81                        Polar Ice           73.33
83                     Polar Sports              80
84                       Polar Wave              60
85                    Polar Extreme            72.5
88                     Edge Extreme              80
91                        Seeker 35           79.19
92                        Seeker 50           92.58
93           TrailChef Double Flame              75
94                   Seeker Extreme           94.12
97                   Glacier Deluxe           56.25
98                      Glacier GPS           78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [103]:
margin_range = (product['MARGIN'].astype(float) < .20) | (product['MARGIN'].astype(float) > .60)
pd.DataFrame(product.loc[margin_range, ['PRODUCT_NAME', 'MARGIN']])

PRODUCT_NAME MARGIN
29                 Hibernator Pad    .17
32   Canyon Mule Climber Backpack    .17
100             BugShield Natural    .67
101               BugShield Spray    .67
102         BugShield Lotion Lite     .7
103              BugShield Lotion    .63
105             BugShield Extreme    .63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [104]:
currency = country['CURRENCY_NAME'] == 'francs'
pd.DataFrame(country.loc[currency, ['COUNTRY']].sort_values(by=['COUNTRY']))

COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [105]:
filter = product['MARGIN'].astype(float) > .50
pd.DataFrame(product.loc[filter, ['INTRODUCTION_DATE']].drop_duplicates())

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [119]:
filter = (sales_branch['ADDRESS2'].notna()) & (sales_branch['REGION'].notna())
pd.DataFrame(sales_branch.loc[filter, ['ADDRESS1', 'CITY']])

ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [107]:
currency = (country['CURRENCY_NAME'] == 'new dollar') | (country['CURRENCY_NAME'] == 'dollars')
pd.DataFrame(country.loc[currency, ['COUNTRY']].sort_values(by=['COUNTRY']))

COUNTRY
6       Australia
15         Canada
3          Taiwan
14  United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [139]:
filter = (retailer_site['POSTAL_ZONE'].str.startswith('D')) & (retailer_site['ADDRESS2'].notna())
pd.DataFrame(retailer_site.loc[filter, ['ADDRESS1', 'ADDRESS2', 'CITY']])

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [118]:
len(returned_item)

706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [135]:
filter = retailer_site[retailer_site['REGION'].notna()]
filter['REGION'].drop_duplicates().count()

110

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [180]:
pd.DataFrame({ 'laagste' : [product['MARGIN'].astype(float).min()], 'hoogste' : [product['MARGIN'].astype(float).max()], 'gemiddelde' : [product['MARGIN'].astype(float).mean()] })

laagste  hoogste  gemiddelde
0     0.17      0.7    0.396609

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [162]:
retailer_site['ADDRESS2'].isna().sum()

286

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [183]:
filter = order_details['UNIT_SALE_PRICE'].astype(float) < order_details['UNIT_PRICE'].astype(float)
filtered = pd.DataFrame(order_details.loc[filter, :])
filtered['UNIT_COST'].astype(float).mean()

125.42305809556548

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [187]:
filter = sales_staff['POSITION_EN'].value_counts().reset_index()
filter.columns = ['POSITION_EN', 'COUNT']
filter

POSITION_EN  COUNT
0  Level 2 Sales Representative     39
1  Level 3 Sales Representative     24
2                Branch Manager     19
3  Level 1 Sales Representative     13
4              District Manager      4
5              Regional Manager      2
6               General Manager      1

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [190]:
phone_counts = sales_staff['WORK_PHONE'].value_counts()
filter = phone_counts > 4
result = phone_counts[filter].reset_index()
result.columns = ['WORK_PHONE', 'COUNT']
result

WORK_PHONE  COUNT
0     1 (305) 557-4810      6
1     1 (617) 268-6754      6
2  +31 (0)20 692 93 94      5
3     1 (403) 232-5986      5
4  +(61) 03 2982  4242      5
5    +(44) 121 3505267      5
6     1 (310) 281-5722      5
7    +33 1 68 94 52 20      5
8     1 (206) 292-0012      5
9     1 (416) 493-5595      5

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [109]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [110]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 